### ISSUE TITLE
speed up button

### ISSUE THREAD

**Issue by antonpiatek** [andybalaam/rabbit-escape#1]
I often get to the point where I have everything in place but I'm bored waiting for all the rabbits to run. A 2x (or more) speed button would be good

**Comment by andybalaam** [comment]
Yes, definitely.

**Comment by colonelfazackerley** [comment]
Given the difficulty escalator we are riding, I suggest slow down buttons too.

**Comment by andybalaam** [comment]
My initial reaction is that that is too much like cheating. When paused you can't place tokens but presumably in slow mode you could?

**Comment by antonpiatek** [comment]
Speed up was primarily to move them faster, often when you're all set up
and just want them to run to the finish. Not sure I see a need for slow
mode except to make it easier, in which case allow assignments while paused

**Comment by colonelfazackerley** [comment]
I totally understand the need for a speed up button.

Have you tried the newer level sets? Or some of the new proposed levels in github issues? They are hard. Copious use of pause and a slow down mode would help.

**Comment by colonelfazackerley** [comment]
Not sure about token drops while paused. Might be considered cheaty. Let's see what The Architect says.

**Comment by andybalaam** [comment]
I think token drops while paused and slow motion are both cheats. If the levels are too hard we should make them easier. On the other hand,  if someone plays through all 100 existing levels they are probably looking for a challenge.

**Comment by colonelfazackerley** [comment]
I was wondering about how to implement this

GeneralPhysics
```java
   @Override
    public long step( long simulation_time, long frame_start_time )
    {
        for ( int skipped = 0; skipped < max_allowed_skips; ++skipped )
        {
            if ( simulation_time >= frame_start_time )
            {
                break;
            }

            frame += n;

            if ( frame == 10 )
            {
                frame = 0;

                doInterpreterActions();
                worldModifier.step();
                checkWon();
                notifyStatsListeners();
            }

            simulation_time += simulation_time_step_ms;
        }

        return simulation_time;
    }
```

Vary n between 1,2,5,10. This would drop frames. Is this a good way to implement speedup? I compiled this and ran it in demo mode. Did not look too bad.

Or should we aim for 20 frames per second when running at 2x, instead of this which would give 10 frames per second, but not display every other frame. 

Could the higher frame rates be achieved with a timing tweak in GameLoop?

**Comment by andybalaam** [comment]
I was expecting to skip frames, probably jumping 5 at a time. I don't think there is any need to have a faster frame rate. I also don't think we need more than 2 speeds: fast and normal.

**Comment by andybalaam** [comment]
I'm not sure where frame normally gets incremented, but I would have expected to have n=4 to do 5x speed?

**Comment by colonelfazackerley** [comment]
We can experiment with 2x and 5x speedup, see which is best. having only normal and fast will simplify the GUI, so I agree that is a good idea. we can have a single button that toggles speeds.

how about a button with a single "play" triangle for normal and multiple triangles for "fast forward"? if we did use this, should it display the current state, or the state you would switch to if you tap/clicked it?

**Comment by andybalaam** [comment]
Could we share it with the pause button somehow? Probably not. How about just a double arrow that is either active looking or not? Do maybe Black when you are at normal speed and glowing white when fast.

**Comment by colonelfazackerley** [comment]
we better do the same for the pause button. always looks like a pause button, but goes glowy when active.

**Comment by colonelfazackerley** [comment]
How about this?

inactive
![speedup-inactive](https://cloud.githubusercontent.com/assets/9596719/12371226/8fcfe6a6-bc24-11e5-86b6-c45c89ba3676.png)

active
![speedup-active](https://cloud.githubusercontent.com/assets/9596719/12371229/a26c0664-bc24-11e5-832a-9cdaa3f7eb2c.png)

If this is approved, I think the pause button should be the same. It is always pause shaped, but goes from looking dull to glowy.

**Comment by andybalaam** [comment]
I think I prefer the black-on-white of the current buttons, but that does raise the question of how we can make something glow white on a quite background.  How about black when inactive and glowing red when active.

I am torn about the pause button because I like the way it works now, but I take your point that we should be consistent.

**Comment by andybalaam** [comment]
While we're at it, I'm not sure the mute button is as clear as it could be either.  It would be nice to be able to click on a "line through loudspeaker" button to mute, so that adds weight to the active/inactive style.

**Comment by colonelfazackerley** [comment]
active/inactive options
![menu_speedup_active](https://cloud.githubusercontent.com/assets/9596719/12372583/a9cbfbf4-bc55-11e5-9441-e2dedf724985.png)
![menu_speedup_inactive](https://cloud.githubusercontent.com/assets/9596719/12372585/b1708c4e-bc55-11e5-991e-e76e96c7d433.png)

**Comment by colonelfazackerley** [comment]
I red looks a bit angry. I prefer green it says "go" to me.
![menu_speedup_active](https://cloud.githubusercontent.com/assets/9596719/12372594/132750f8-bc56-11e5-8c14-b01c63cd6c6c.png)

**Comment by colonelfazackerley** [comment]
I think this is a slightly better glow effect.
![menu_speedup_active](https://cloud.githubusercontent.com/assets/9596719/12372863/9ba2b268-bc5d-11e5-9b84-75a654df242b.png)

I can do similar for mute and pause, if this is approved.

**Comment by tttppp** [comment]
I like the latest one!

**Comment by andybalaam** [comment]
Yes very cool. Please do similar for pause and mute!

**Comment by andybalaam** [comment]
Implemented in #291 and #315.

Issue #1 finally closed!

This feature will be in release 0.8.
